# Project Description
A semiconductor manufacturer has 437 sensors tracking their process and whether. The sensor outputs is associated with a pass or fail value. Using the data from these sensors build a model that can predict whether a semiconductor should pass or fail.

# Data Preprocessing

## Importing the Libraries

In [209]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

## Importing the Dataset

In [210]:
dataset = pd.read_csv('SemiconductorManufacturingProcessDataset.csv')

## Showing the Dataset in a Table

In [211]:
pd.DataFrame(dataset)

Time  Sensor 1  Sensor 2   Sensor 3   Sensor 4  Sensor 5  \
0      7/19/2008 11:55   3030.93   2564.00  2187.7333  1411.1265    1.3602   
1      7/19/2008 12:32   3095.78   2465.14  2230.4222  1463.6606    0.8294   
2      7/19/2008 13:17   2932.61   2559.94  2186.4111  1698.0172    1.5102   
3      7/19/2008 14:43   2988.72   2479.90  2199.0333   909.7926    1.3204   
4      7/19/2008 15:22   3032.24   2502.87  2233.3667  1326.5200    1.5334   
...                ...       ...       ...        ...        ...       ...   
1562  10/16/2008 15:13   2899.41   2464.36  2179.7333  3085.3781    1.4843   
1563  10/16/2008 20:49   3052.31   2522.55  2198.5667  1124.6595    0.8763   
1564   10/17/2008 5:26   2978.81   2379.78  2206.3000  1110.4967    0.8236   
1565   10/17/2008 6:01   2894.92   2532.01  2177.0333  1183.7287    1.5726   
1566   10/17/2008 6:07   2944.92   2450.76  2195.4444  2914.1792    1.5978   

      Sensor 6  Sensor 7  Sensor 8  Sensor 9  ...  Sensor 429  Sensor 430  \
0      97.6133    0.1242    1.5005    0.0162  ...     14.9509      0.5005   
1     102.3433    0.1247    1.4966   -0.0005  ...     10.9003      0.5019   
2      95.4878    0.1241    1.4436    0.0041  ...      9.2721      0.4958   
3     104.2367    0.1217    1.4882   -0.0124  ...      8.5831      0.4990   
4     100.3967    0.1235    1.5031   -0.0031  ...     10.9698      0.4800   
...        ...       ...       ...       ...  ...         ...         ...   
1562   82.2467    0.1248    1.3424   -0.0045  ...     11.7256      0.4988   
1563   98.4689    0.1205    1.4333   -0.0061  ...     17.8379      0.4975   
1564   99.4122    0.1208       NaN       NaN  ...     17.7267      0.4987   
1565   98.7978    0.1213    1.4622   -0.0072  ...     19.2104      0.5004   
1566   85.1011    0.1235       NaN       NaN  ...     22.9183      0.4987   

      Sensor 431  Sensor 432  Sensor 433  Sensor 434  Sensor 435  Sensor 436  \
0         0.0118      0.0035      2.3630         NaN         NaN         NaN   
1         0.0223      0.0055      4.4447      0.0096      0.0201      0.0060   
2         0.0157      0.0039      3.1745      0.0584      0.0484      0.0148   
3         0.0103      0.0025      2.0544      0.0202      0.0149      0.0044   
4         0.4766      0.1045     99.3032      0.0202      0.0149      0.0044   
...          ...         ...         ...         ...         ...         ...   
1562      0.0143      0.0039      2.8669      0.0068      0.0138      0.0047   
1563      0.0131      0.0036      2.6238      0.0068      0.0138      0.0047   
1564      0.0153      0.0041      3.0590      0.0197      0.0086      0.0025   
1565      0.0178      0.0038      3.5662      0.0262      0.0245      0.0075   
1566      0.0181      0.0040      3.6275      0.0117      0.0162      0.0045   

      Sensor 437  Pass/Fail  
0            NaN       Pass  
1       208.2045       Pass  
2        82.8602       Fail  
3        73.8432       Pass  
4        73.8432       Pass  
...          ...        ...  
1562    203.1720       Pass  
1563    203.1720       Pass  
1564     43.5231       Pass  
1565     93.4941       Pass  
1566    137.7844       Pass  

[1567 rows x 439 columns]

## Seperate The Input and Output
Here, we put the independent variables in X and the dependent variable in y. 

In [212]:
X = dataset.iloc[:, :-1].values
Y = dataset.iloc[:, -1].values

## Showing the Input Data in a Table format

In [213]:
pd.DataFrame(X)

0        1        2        3        4       5        6    \
0      7/19/2008 11:55  3030.93     2564  2187.73  1411.13  1.3602  97.6133   
1      7/19/2008 12:32  3095.78  2465.14  2230.42  1463.66  0.8294  102.343   
2      7/19/2008 13:17  2932.61  2559.94  2186.41  1698.02  1.5102  95.4878   
3      7/19/2008 14:43  2988.72   2479.9  2199.03  909.793  1.3204  104.237   
4      7/19/2008 15:22  3032.24  2502.87  2233.37  1326.52  1.5334  100.397   
...                ...      ...      ...      ...      ...     ...      ...   
1562  10/16/2008 15:13  2899.41  2464.36  2179.73  3085.38  1.4843  82.2467   
1563  10/16/2008 20:49  3052.31  2522.55  2198.57  1124.66  0.8763  98.4689   
1564   10/17/2008 5:26  2978.81  2379.78   2206.3   1110.5  0.8236  99.4122   
1565   10/17/2008 6:01  2894.92  2532.01  2177.03  1183.73  1.5726  98.7978   
1566   10/17/2008 6:07  2944.92  2450.76  2195.44  2914.18  1.5978  85.1011   

         7       8       9    ...     428      429     430     431     432  \
0     0.1242  1.5005  0.0162  ...  1.6765  14.9509  0.5005  0.0118  0.0035   
1     0.1247  1.4966 -0.0005  ...  1.1065  10.9003  0.5019  0.0223  0.0055   
2     0.1241  1.4436  0.0041  ...  2.0952   9.2721  0.4958  0.0157  0.0039   
3     0.1217  1.4882 -0.0124  ...  1.7585   8.5831   0.499  0.0103  0.0025   
4     0.1235  1.5031 -0.0031  ...  1.6597  10.9698    0.48  0.4766  0.1045   
...      ...     ...     ...  ...     ...      ...     ...     ...     ...   
1562  0.1248  1.3424 -0.0045  ...  1.4879  11.7256  0.4988  0.0143  0.0039   
1563  0.1205  1.4333 -0.0061  ...  1.0187  17.8379  0.4975  0.0131  0.0036   
1564  0.1208     NaN     NaN  ...  1.2237  17.7267  0.4987  0.0153  0.0041   
1565  0.1213  1.4622 -0.0072  ...  1.7085  19.2104  0.5004  0.0178  0.0038   
1566  0.1235     NaN     NaN  ...  1.2878  22.9183  0.4987  0.0181   0.004   

          433     434     435     436      437  
0       2.363     NaN     NaN     NaN      NaN  
1      4.4447  0.0096  0.0201   0.006  208.204  
2      3.1745  0.0584  0.0484  0.0148  82.8602  
3      2.0544  0.0202  0.0149  0.0044  73.8432  
4     99.3032  0.0202  0.0149  0.0044  73.8432  
...       ...     ...     ...     ...      ...  
1562   2.8669  0.0068  0.0138  0.0047  203.172  
1563   2.6238  0.0068  0.0138  0.0047  203.172  
1564    3.059  0.0197  0.0086  0.0025  43.5231  
1565   3.5662  0.0262  0.0245  0.0075  93.4941  
1566   3.6275  0.0117  0.0162  0.0045  137.784  

[1567 rows x 438 columns]

## A Quick Check of the Output Data

In [214]:
pd.DataFrame(Y)

0
0     Pass
1     Pass
2     Fail
3     Pass
4     Pass
...    ...
1562  Pass
1563  Pass
1564  Pass
1565  Pass
1566  Pass

[1567 rows x 1 columns]

## Taking care of missing data

In [215]:
from sklearn.impute import SimpleImputer
imputer = SimpleImputer(missing_values=np.nan, strategy='mean')
imputer.fit(X[:, 1:])
X[:, 1:] = imputer.transform(X[:, 1:])

In [216]:
# A quick check
pd.DataFrame(X)

0        1        2        3        4       5        6    \
0      7/19/2008 11:55  3030.93     2564  2187.73  1411.13  1.3602  97.6133   
1      7/19/2008 12:32  3095.78  2465.14  2230.42  1463.66  0.8294  102.343   
2      7/19/2008 13:17  2932.61  2559.94  2186.41  1698.02  1.5102  95.4878   
3      7/19/2008 14:43  2988.72   2479.9  2199.03  909.793  1.3204  104.237   
4      7/19/2008 15:22  3032.24  2502.87  2233.37  1326.52  1.5334  100.397   
...                ...      ...      ...      ...      ...     ...      ...   
1562  10/16/2008 15:13  2899.41  2464.36  2179.73  3085.38  1.4843  82.2467   
1563  10/16/2008 20:49  3052.31  2522.55  2198.57  1124.66  0.8763  98.4689   
1564   10/17/2008 5:26  2978.81  2379.78   2206.3   1110.5  0.8236  99.4122   
1565   10/17/2008 6:01  2894.92  2532.01  2177.03  1183.73  1.5726  98.7978   
1566   10/17/2008 6:07  2944.92  2450.76  2195.44  2914.18  1.5978  85.1011   

         7        8            9    ...     428      429     430     431  \
0     0.1242   1.5005       0.0162  ...  1.6765  14.9509  0.5005  0.0118   
1     0.1247   1.4966      -0.0005  ...  1.1065  10.9003  0.5019  0.0223   
2     0.1241   1.4436       0.0041  ...  2.0952   9.2721  0.4958  0.0157   
3     0.1217   1.4882      -0.0124  ...  1.7585   8.5831   0.499  0.0103   
4     0.1235   1.5031      -0.0031  ...  1.6597  10.9698    0.48  0.4766   
...      ...      ...          ...  ...     ...      ...     ...     ...   
1562  0.1248   1.3424      -0.0045  ...  1.4879  11.7256  0.4988  0.0143   
1563  0.1205   1.4333      -0.0061  ...  1.0187  17.8379  0.4975  0.0131   
1564  0.1208  1.46286 -0.000841022  ...  1.2237  17.7267  0.4987  0.0153   
1565  0.1213   1.4622      -0.0072  ...  1.7085  19.2104  0.5004  0.0178   
1566  0.1235  1.46286 -0.000841022  ...  1.2878  22.9183  0.4987  0.0181   

         432      433        434        435         436      437  
0     0.0035    2.363  0.0214584  0.0164749  0.00528333  99.6701  
1     0.0055   4.4447     0.0096     0.0201       0.006  208.204  
2     0.0039   3.1745     0.0584     0.0484      0.0148  82.8602  
3     0.0025   2.0544     0.0202     0.0149      0.0044  73.8432  
4     0.1045  99.3032     0.0202     0.0149      0.0044  73.8432  
...      ...      ...        ...        ...         ...      ...  
1562  0.0039   2.8669     0.0068     0.0138      0.0047  203.172  
1563  0.0036   2.6238     0.0068     0.0138      0.0047  203.172  
1564  0.0041    3.059     0.0197     0.0086      0.0025  43.5231  
1565  0.0038   3.5662     0.0262     0.0245      0.0075  93.4941  
1566   0.004   3.6275     0.0117     0.0162      0.0045  137.784  

[1567 rows x 438 columns]

## Encoding Categorical Data

### Encoding the Dependent Variable

In [217]:
from sklearn.preprocessing import LabelEncoder
LaEnc = LabelEncoder()
Y = LaEnc.fit_transform(Y)

In [218]:
# a qucik check
pd.DataFrame(Y)

0
0     1
1     1
2     0
3     1
4     1
...  ..
1562  1
1563  1
1564  1
1565  1
1566  1

[1567 rows x 1 columns]

## Feature Scaling

In [219]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X[:, 1:] = sc.fit_transform(X[:, 1:])

In [220]:
# Quick Check
pd.DataFrame(X)

0         1          2          3         4          5    \
0      7/19/2008 11:55  0.224309   0.849725  -0.436273  0.033555 -0.0505805   
1      7/19/2008 12:32   1.10714   -0.38291    1.01714  0.153067 -0.0600446   
2      7/19/2008 13:17  -1.11416   0.799102  -0.481289  0.686213  -0.047906   
3      7/19/2008 14:43 -0.350312  -0.198875 -0.0515471  -1.10695 -0.0512901   
4      7/19/2008 15:22  0.242143  0.0875259    1.11739 -0.158919 -0.0474923   
...                ...       ...        ...        ...       ...        ...   
1562  10/16/2008 15:13  -1.56612  -0.392636  -0.708645   3.84237 -0.0483678   
1563  10/16/2008 20:49  0.515363   0.332906 -0.0674332 -0.618139 -0.0592084   
1564   10/17/2008 5:26  -0.48522   -1.44722   0.195859 -0.650359 -0.0601481   
1565   10/17/2008 6:01  -1.62725   0.450858  -0.800571 -0.483761 -0.0467934   
1566   10/17/2008 6:07 -0.946577  -0.562207  -0.173737   3.45291  -0.046344   

           6          7           8            9    ...       428        429  \
0     -0.56379   0.266269    0.509826      1.12842  ... -0.229797   -0.13552   
1     0.198217   0.322244    0.456999    0.0225817  ... -0.263493  -0.460054   
2     -0.90621   0.255074   -0.260907     0.327183  ... -0.205046  -0.590505   
3     0.503246  -0.013602    0.343218    -0.765408  ...  -0.22495  -0.645708   
4    -0.115382   0.187905    0.545044    -0.149584  ... -0.230791  -0.454486   
...        ...        ...         ...          ...  ...       ...        ...   
1562  -3.03936   0.333438     -1.6317    -0.242289  ... -0.240947  -0.393931   
1563 -0.425952   -0.14794   -0.400425    -0.348237  ... -0.268684  0.0957864   
1564 -0.273986  -0.114356           0 -7.17933e-18  ... -0.256565  0.0868771   
1565 -0.372966 -0.0583814 -0.00896247    -0.421077  ... -0.227906   0.205751   
1566  -2.57952   0.187905           0 -7.17933e-18  ... -0.252776   0.502828   

            430         431        432        433       434        435  \
0      0.118699    -0.20489 -0.0932074  -0.197113         0          0   
1      0.530203    0.406679   0.444706   0.385059 -0.960174   0.411853   
2      -1.26278   0.0222639  0.0143753  0.0298325   2.99115    3.62706   
3     -0.322199   -0.292257  -0.362164  -0.283417 -0.101895  -0.178927   
4       -5.9069     26.8672    27.0714    26.9133 -0.101895  -0.178927   
...         ...         ...        ...        ...       ...        ...   
1562  -0.380985  -0.0592786  0.0143753 -0.0561914  -1.18689    -0.3039   
1563  -0.763096   -0.129172 -0.0663118  -0.124177  -1.18689    -0.3039   
1564  -0.410378 -0.00103397  0.0681666 -0.0024684  -0.14238  -0.894681   
1565  0.0893057    0.144578 -0.0125204   0.139376  0.383924   0.911745   
1566  -0.410378    0.162051  0.0412709   0.156519 -0.790138 -0.0312323   

              436         437  
0     6.05437e-16  1.5145e-16  
1        0.250124     1.15669  
2         3.32142   -0.179149  
3       -0.308293   -0.275246  
4       -0.308293   -0.275246  
...           ...         ...  
1562     -0.20359     1.10306  
1563     -0.20359     1.10306  
1564    -0.971413   -0.598378  
1565      0.77364  -0.0658194  
1566    -0.273392    0.406198  

[1567 rows x 438 columns]

## Remove Specific Columns
Remove column containing date. 

In [221]:
X = np.delete(X, 0, 1)
X

array([[0.22430902093636199, 0.8497245455534013, -0.436273028037545, ...,
        0.0, 6.05437138808527e-16, 1.5145001106248855e-16],
       [1.1071364666283177, -0.38291012836538457, 1.0171365185603627,
        ..., 0.41185317333833354, 0.25012436976549723,
        1.1566891300187565],
       [-1.114158125716411, 0.7991024863170044, -0.4812893695809402, ...,
        3.627063334759813, 3.32141895665346, -0.17914858010239992],
       ...,
       [-0.48522014590194756, -1.4472202210793945, 0.1958589554246609,
        ..., -0.8946809911968965, -0.9714132500194879,
        -0.5983777080867355],
       [-1.6272462741841007, 0.45085763053558175, -0.8005710574690204,
        ..., 0.9117445058561608, 0.7736404925304907,
        -0.06581941694414414],
       [-0.9465774648610855, -0.5622069785672021, -0.17373671637875118,
        ..., -0.03123232593883151, -0.27339175299949653,
        0.40619768261745587]], dtype=object)

## Splitting the Dataset into the Training set and Test set

In [222]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 1)

# Classification Model Analysis
## KNN
Train the regression model.

In [223]:
from sklearn.neighbors import KNeighborsClassifier 

# instantiate knn object
knn = KNeighborsClassifier(n_neighbors=3)
# train model
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

Print out the models accuracy.

In [224]:
knn.score(X_test, y_test)

0.9331210191082803

# Conclusion
The models accuracy using KNN with 3 neighbors is approximately 93%. This could potentially be improved by reducing the number of sensors or by using a different classification algorithm.